File: code/3sum.java
public class Solution {
    /*
     * @param numbers: Give an array numbers of n integer
     * @return: Find all unique triplets in the array which gives the sum of zero.
     */
    public List<List<Integer>> threeSum(int[] numbers) {
        // write your code here

        List<List<Integer>> ans = new ArrayList<>();
        int len = numbers.length;
        if (len <= 0) return ans;
        Arrays.sort(numbers);
        for (int i = 0; i < len - 2; i++) {
            twoSum(numbers, ans, i, len - 1);
            while (i < len - 2 && numbers[i] == numbers[i + 1])
                i++;
        }
        return ans;
    }

    private void twoSum(int[] numbers, List<List<Integer>> ans, int start, int end) {
        int target = -numbers[start];

        int i = start + 1;
        int j = end;
        while (i < j) {
            int sum = numbers[i] + numbers[j];
            if (sum == target) {
                List<Integer> curr = new ArrayList<>();
                curr.add(numbers[start]);
                curr.add(numbers[i]);
                curr.add(numbers[j]);
                ans.add(curr);

                while (i < j && numbers[i] == numbers[i + 1]) i++;
                while (i < j && numbers[j] == numbers[j - 1]) j--;
                i++;
                j--;
            } else if (sum < target) {
                i++;
            } else {
                j--;
            }
        }
    }


}

File: code/medium1/3sum.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        
        sort(nums.begin(), nums.end());

        vector<vector<int>> ans;

        for (int i = 0; i < nums.size(); i++) {
            int target = -nums[i];
            int start = i + 1, end = nums.size() - 1;

            while (start < end) {
                int sum = nums[start] + nums[end];
                if (sum > target) {
                    end--;
                } else if (sum < target) {
                    start++;
                } else {
                    vector<int> temp = {nums[i], nums[start], nums[end]};
                    ans.push_back(temp);
                    while (start + 1 < nums.size() && nums[start + 1] == nums[start]) start++;
                    start++;
                    while (end - 1 >= i && nums[end - 1] == nums[end]) end--;
                    end--;
                }
            }

            while (i + 1 < nums.size() && nums[i + 1] == nums[i]) i++;
        }

        return ans;
    }
};